### ISSUE TITLE
Behavior of filter_by_attrs() does not match netCDF4.Dataset.get_variables_by_attributes

### ISSUE THREAD

**Issue by DocOtak** [pydata/xarray#2315]
When using the `filter_by_attrs()` method of a Dataset it was returning more matches than I expected. The returned match set appears to be the logical OR of all the key=value pairs passed into the method. This does not match the behavior of the `netCDF4.Dataset.get_variables_by_attributes` which returns the logical AND of all the key=value pairs passed into it. I couldn't tell from the documentation if this was the intended behavior.

#### Minimal Example
```python
import xarray as xr
example_dataset = xr.Dataset({
    "var1": xr.DataArray([], attrs={"standard_name": "example1", "priority": 0}),
    "var2": xr.DataArray([], attrs={"standard_name": "example2"})
})
example_dataset.filter_by_attrs(standard_name="example2", priority=0)
```

#### Example Output
```python
<xarray.Dataset>
Dimensions:  (dim_0: 0)
Dimensions without coordinates: dim_0
Data variables:
    var1     (dim_0) float64 
    var2     (dim_0) float64
```

#### Expected Output

```python
<xarray.Dataset>
Dimensions:  ()
Data variables:
    *empty*
```

Alternatively, chaining calls to filter_by_attrs will result in the expected behavior:
```python
import xarray as xr
example_dataset = xr.Dataset({
    "var1": xr.DataArray([], attrs={"standard_name": "example1", "priority": 0}),
    "var2": xr.DataArray([], attrs={"standard_name": "example2"})
})
example_dataset.filter_by_attrs(standard_name="example2").filter_by_attrs(priority=0)

<xarray.Dataset>
Dimensions:  ()
Data variables:
    *empty*
```

#### Output of ``xr.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Darwin
OS-release: 17.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.8
pandas: 0.23.3
numpy: 1.14.5
scipy: 1.1.0
netCDF4: 1.4.0
h5netcdf: None
h5py: None
Nio: None
zarr: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.18.2
distributed: 1.22.0
matplotlib: 2.2.2
cartopy: 0.16.0
seaborn: None
setuptools: 39.0.1
pip: 10.0.1
conda: None
pytest: 3.6.3
IPython: 6.4.0
sphinx: None

</details>

**Comment by shoyer** [comment]
cc @ocefpaf, who contributed this method originally

**Comment by ocefpaf** [comment]
> it appears that this scenario was not directly contemplated.

Correct. I did not foresee that use case. Are you sure that `netCDF4.Dataset.get_variables_by_attributes` behaves as a logical AND? The code is virtually a copy-n-paste from there.

**Comment by DocOtak** [comment]
I'm fairly certain that the `netCDF4.Dataset.get_variables_by_attributes` behaves as a logical AND.

Here is the currently implementation body from https://github.com/Unidata/netcdf4-python/blob/master/netCDF4/_netCDF4.pyx#L2868
```python
vs = []

has_value_flag  = False
# this is a hack to make inheritance work in MFDataset
# (which stores variables in _vars)
_vars = self.variables
if _vars is None: _vars = self._vars
for vname in _vars:
   var = _vars[vname]
    for k, v in kwargs.items():
       if callable(v):
           has_value_flag = v(getattr(var, k, None))
           if has_value_flag is False:
               break
       elif hasattr(var, k) and getattr(var, k) == v:
           has_value_flag = True
       else:
           has_value_flag = False
           break

   if has_value_flag is True:
        vs.append(_vars[vname])

return vs
```

The difference appears to be in the presence of that `has_value_flag` and the breaks in the innermost loop. I must admit I had a little trouble following the above code, but it seems that when any of they key=value tests fails, it will stop checking the current variable (DataArray in the xarray context) and check the next variable.

**Comment by ocefpaf** [comment]
Got it. I cannot dig into this at the moment but having both implementations working in a consistent way would be nice. Do you want to send a PR?

**Comment by DocOtak** [comment]
I can work on a PR tomorrow. Does the benefit of having the same behavior as the netCDF4 library warrant a potentially breaking change for existing code which relies on the current behavior of `filter_by_attrs()`? This might need adding a new method with the same behavior as netCDF4 and keeping the existing one as is (with appropriate documentation updates).

**Comment by ocefpaf** [comment]
> I can work on a PR tomorrow. Does the benefit of having the same behavior as the netCDF4 library warrant a potentially breaking change for existing code which relies on the current behavior of filter_by_attrs()?

IMO, yes.

> This might need adding a new method with the same behavior as netCDF4 and keeping the existing one as is (with appropriate documentation updates).

That is up to xarray devs but I personally don't think it is necessary.

**Comment by shoyer** [comment]
I agree, I think we can change the behavior here to something more rationale. Arguably the existing behavior is a bug.
